deployment/roles/toy.database/tasks/main.yml

```diff
@@ -0,0 +1,27 @@
+---
+- name: Install PostgreSQL
+  apt: pkg="{{ item }}" state=present
+  with_items:
+    - postgresql
+    - libpq-dev
+    - postgresql-client
+    - postgresql-client-common
+
+- name: Add UbuntuGIS apt repository
+  apt_repository:
+    repo: ppa:ubuntugis/ppa
+
+- name: Update APT cache
+  apt: update_cache=yes
+
+- name: Install PostGIS
+  apt: pkg="{{ item }}" state=present
+  with_items:
+    - postgis
+    - postgresql-10-postgis-2.4-scripts
+
+- name: Install PostgreSQL driver for Python
+  import_tasks: psycopg2.yml
```

Comment:
When we split things apart, we usually do it via Roles rather than importing tasks like this; using roles has the advantage of providing a built-in project structure and making reuse across projects somewhat easier. Not critical here though. https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#roles

Although, given that there's just one task in the imported task list, I wonder if that task could be moved in here?

Reply:
I moved the psycopg2 installation step into this file `toy.database/tasks/main`, but am still little bit confused as to what the ideal scoping for a Role should be. Based on the cac-tripplanner repo, it seems like the psycopg2 installation step is [within the database role](https://github.com/azavea/cac-tripplanner/blob/681fa2dc95ea8f6a330844babed122254db7ab5e/deployment/ansible/roles/cac-tripplanner.database/tasks/main.yml#L22) - is this something that you feel might make sense to move into a separate role?

Or, to look at this one level higher - are there any suggestions you have on the overall structure of tasks here? Or any general guidelines to follow re: what constitutes a separate role?

Reply:
Yeah, the proper place to separate things is always a bit of a judgment call. I don't have any grand theories about when to break a set of tasks out into a role, but generally I would say that if it's just a single task, it probably doesn't need a separate role. Likewise, if it's a bunch of unrelated tasks then that probably wouldn't be a good match for a single Role. Here are probably the two most common things that we've used roles for in the past:
- Installing and configuring a service, like Postgres or Redis
- Installing a library that has a complicated build process

Splitting out into roles also becomes more important as the project grows because there's a higher likelihood that it will be useful to set up the same dependencies on multiple machine types.

Reply:
Got it, thanks for the clarification!